src/Identity/Extensions.Core/src/UserManager.cs

```diff
@@ -2276,7 +2277,60 @@ public virtual string GenerateNewAuthenticatorKey()
     /// </summary>
     /// <returns></returns>
     protected virtual string CreateTwoFactorRecoveryCode()
-        => Guid.NewGuid().ToString().Substring(0, 8);
+    {
+        var recoveryCode = new StringBuilder(11);
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append('-');
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        recoveryCode.Append(GetRandomRecoveryCodeChar());
+        return recoveryCode.ToString();
+    }
+
+    private static char GetRandomRecoveryCodeChar()
+    {
+        // We don't want to use any confusing characters like 0/O 1/I/L/l
+        // Taken from windows valid product key source
+        const string AllowedChars = "23456789BCDFGHJKMNPQRTVWXY";
```

Comment:
Just FYI https://vcsjones.dev/csharp-readonly-span-bytes-static/

Reply:
Yeah, totally agree that the lookup table could be faster, but yeah this is done generally only on user creation, so readability is more important.  That said CharArray change seems fine though, do you want me to make that?

Reply:
I'd prefer personally the char array, more common for the usage. And also as a private static field  in that case.